src/radiation/mocmc.cpp

```diff
@@ -0,0 +1,947 @@
+            }
+            Real JBB = opac_d.EnergyDensityFromTemperature(v(iblock, pT, k, j, i),
+                                                           dev_species[ispec]);
+
+            ClosureMOCMC<Vec, Tens2> c(con_v, &g);
+
+            Real dE = 0;
+            Vec cov_dF;
+
+            // Calculate Inu0
+            for (int bin = 0; bin < nu_bins; bin++) {
+              v(iblock, Inu0(ispec, bin), k, j, i) = 0.;
+            }
+            for (int n = 0; n < nsamp; n++) {
+              const int nswarm = swarm_d.GetFullIndex(k, j, i, n);
+              const Real dOmega = (mu_hi(nswarm) - mu_lo(nswarm)) *
+                                  (phi_hi(nswarm) - phi_lo(nswarm)) / (4. * M_PI);
+              for (int bin = 0; bin < nu_bins; bin++) {
+                v(iblock, Inu0(ispec, bin), k, j, i) +=
+                    Inuinv(bin, ispec, nswarm) * pow(nusamp(bin), 3) * dOmega;
+              }
+            }
+
+            // Frequency-average opacities
+            Real kappaJ = 0.;
+            Real kappaH = 0.;
+
+            Real Itot = 0.;
+            for (int bin = 0; bin < nu_bins; bin++) {
+              kappaJ += opac_d.AngleAveragedAbsorptionCoefficient(
+                            v(iblock, pdens, k, j, i), v(iblock, pT, k, j, i),
+                            v(iblock, pye, k, j, i), dev_species[ispec], nusamp(bin)) *
+                        v(iblock, Inu0(ispec, bin), k, j, i) * nusamp(bin);
+              Itot += v(iblock, Inu0(ispec, bin), k, j, i) * nusamp(bin);
+            }
+
+            // Trapezoidal rule
+            kappaJ -= 0.5 *
+                      opac_d.AngleAveragedAbsorptionCoefficient(
+                          v(iblock, pdens, k, j, i), v(iblock, pT, k, j, i),
+                          v(iblock, pye, k, j, i), dev_species[ispec], nusamp(0)) *
+                      v(iblock, Inu0(ispec, 0), k, j, i) * nusamp(0);
+            kappaJ -=
+                0.5 *
+                opac_d.AngleAveragedAbsorptionCoefficient(
+                    v(iblock, pdens, k, j, i), v(iblock, pT, k, j, i),
+                    v(iblock, pye, k, j, i), dev_species[ispec], nusamp(nu_bins - 1)) *
+                v(iblock, Inu0(ispec, nu_bins - 1), k, j, i) * nusamp(nu_bins - 1);
+            Itot -= 0.5 *
+                    (v(iblock, Inu0(ispec, 0), k, j, i) * nusamp(0) +
+                     v(iblock, Inu0(ispec, nu_bins - 1), k, j, i) * nusamp(nu_bins - 1));
+            kappaJ = robust::ratio(kappaJ, Itot);
+            kappaH = kappaJ;
+            kappaJ *= (1. - scattering_fraction);
+
+            Real tauJ = alpha * dt * kappaJ;
+            Real tauH = alpha * dt * kappaH;
+
+            // Store kappaH for asymptotic fluxes
+            v(iblock, idx_kappaH(ispec), k, j, i) = kappaH;
+
+            auto status = c.LinearSourceUpdate(Estar, cov_Fstar, con_tilPi, JBB, tauJ,
+                                               tauH, &dE, &cov_dF);
+
```

Comment:
I had initially thought we would have separate tasks for the MOCMC opacity calculation/MOCMC update and the source update for the moments and have fields for kappaJ and kappaH.  The advantage I see there is no duplicated code and maybe more possibility to interleave work, at the cost of more storage.

Reply:
I mentioned above that I wondered how to fold a separate opacity calculation task together with an iterative update task -- maybe the solution is to have a function that calculates opacities, a task that wraps that function, and then a fluid source task that can call the opacity function additional times if necessary?